### PR TITLE
Add function to remove excess spikes

### PR DESCRIPTION
**Pull request by alejoe91**
Sometimes KS returns spikes outside the number of samples. 

This PR:
- asserts that this is not the case when extracting waveforms
- implements a curation function `remove_excess_spikes` that filters out these extra spikes

### CODE REVIEW

File: spikeinterface/core/waveform_extractor.py [hunk 1]
```diff
@@ -245,6 +245,11 @@ def create(cls, recording, sorting, folder, mode="folder", remove_if_exists=Fals
         properties_to_attrs = deepcopy(recording._properties)
         if 'contact_vector' in properties_to_attrs:
             del properties_to_attrs['contact_vector']
+        if has_exceeding_spikes(recording, sorting):
```
Comment (samuelgarcia) [inline spikeinterface/core/waveform_extractor.py:248]: This is a high cost because it does a to vector.
But Iplan to refactor waveforms extractor using only vector version.
So lets go for this for now.

File: spikeinterface/core/waveform_tools.py [hunk 2]
```diff
@@ -351,4 +351,29 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
...
+        The sorting object
+
+    Returns
+    -------
+    bool
+        True if exceeding spikes, False otherwise
+    """
+    spike_vector = sorting.to_spike_vector()
+    for segment_index in range(recording.get_num_segments()):
+        spike_vector_seg = spike_vector[spike_vector["segment_ind"] == segment_index]
```
Comment (samuelgarcia) [inline spikeinterface/core/waveform_tools.py:375]: I would do a serachsorter to be faster.

File: spikeinterface/core/waveform_tools.py [hunk 3]
```diff
@@ -351,4 +351,29 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
...
+        if len(spike_vector_seg) > 0:
+            if np.max(spike_vector_seg["sample_ind"]) > recording.get_num_samples(segment_index=segment_index) - 1:
```
Comment (samuelgarcia) [inline spikeinterface/core/waveform_tools.py:377]: Instead of max I would spike_vector_seg["sample_ind"][-1] which is necesaarly the max.

File: spikeinterface/curation/remove_excess_spikes.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    def __init__(self, parent_segment: BaseSortingSegment, num_samples: int) -> None:
+        super().__init__()
+        self._parent_segment = parent_segment
+        self._num_samples = num_samples
+
+    def get_unit_spike_train(self, unit_id, start_frame: Optional[int] = None,
+                             end_frame: Optional[int] = None) -> np.ndarray:
+        spike_train = self._parent_segment.get_unit_spike_train(unit_id, start_frame=start_frame, end_frame=end_frame)
+
+        return spike_train[spike_train < self._num_samples - 1]
```
Comment (samuelgarcia) [inline spikeinterface/curation/remove_excess_spikes.py:57]: spike_train[spike_train <= self._num_samples - 1]
or
spike_train[spike_train < self._num_samples]
but here you also remove thelast possible spike no ?

File: spikeinterface/curation/remove_excess_spikes.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    Returns
+    -------
+    sorting_without_excess_spikes: Sorting
+        The sorting without any excess spikes.
+    """
+    if has_exceeding_spikes(recording=recording, sorting=sorting):
+        return RemoveExcessSpikesSorting(sorting=sorting, recording=recording)
+    else:
+        return sorting
```
Comment (samuelgarcia) [inline spikeinterface/curation/remove_excess_spikes.py:80]: I understand the why but no sure this is so good to have th hidden stuff.
Anyway it is a detail, lets keep this.

File: spikeinterface/curation/tests/test_remove_excess_spikes.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    num_spikes = 100
+    num_excess_spikes_per_segment = 5
+    times = []
+    labels = []
+    for segment_index in range(recording.get_num_segments()):
+        num_samples = recording.get_num_samples(segment_index=segment_index)
+        times_segment = np.array([], dtype=int)
+        labels_segment = np.array([], dtype=int)
+        for unit in range(num_units):
+            spike_times = np.random.randint(0, num_samples - 1, num_spikes)
```
Comment (samuelgarcia) [inline spikeinterface/curation/tests/test_remove_excess_spikes.py:23]: I think that it should be 
spike_times = np.random.randint(0, num_samples, num_spikes) because the high is not included

File: spikeinterface/curation/tests/test_remove_excess_spikes.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            excess_spikes = np.random.randint(num_samples, num_samples + 100, num_excess_spikes_per_segment)
```
Comment (samuelgarcia) [inline spikeinterface/curation/tests/test_remove_excess_spikes.py:24]: Instead of random I would force hard coded corner case like exactly "num_samples" with a manul list.

File: spikeinterface/core/waveform_extractor.py (hunk 1 above)
Comment (alejoe91) [reply spikeinterface/core/waveform_extractor.py:248]: it takes 1ms

File: spikeinterface/curation/remove_excess_spikes.py (hunk 4 above)
Comment (alejoe91) [reply spikeinterface/curation/remove_excess_spikes.py:57]: oh right!

File: spikeinterface/core/waveform_extractor.py (hunk 1 above)
Comment (samuelgarcia) [reply spikeinterface/core/waveform_extractor.py:248]: I do not believe you of course.
For enormous sorting also ? Or only for a small 10s test file ?

**Comment by alejoe91** [comment]
Thanks @samuelgarcia 

Applied all suggestions

File: spikeinterface/core/waveform_extractor.py (hunk 1 above)
Comment (alejoe91) [reply spikeinterface/core/waveform_extractor.py:248]: >1h long neuropixels output ;) 

I think we have to have an internal caching of the spike vector in the sorting object, but that's another story ;)

**Review by samuelgarcia — APPROVED** [review]
